app/src/ui/blank-slate/create-tutorial-repository-dialog.tsx

```diff
@@ -0,0 +1,280 @@
+    apiRepository: IAPIRepository
+  ) => void
+
+  /**
+   * Event triggered when the component encounters and error while
+   * attempting to create the tutorial repository. Consumers are
+   * intended to display an error message to the end user in response
+   * to this event.
+   */
+  readonly onError: (error: Error) => void
+}
+
+interface ICreateTutorialRepositoryDialogState {
+  /**
+   * Whether or not the dialog is currently in the process of creating
+   * the tutorial repository. When true this will render a spinning
+   * progress icon in the dialog header (if the dialog has a header) as
+   * well as temporarily disable dismissal of the dialog.
+   */
+  readonly loading: boolean
+
+  /**
+   * The current progress in creating the tutorial repository. Undefined
+   * until the creation process starts.
+   */
+  readonly progress?: Progress
+}
+
+const nl = __WIN32__ ? '\r\n' : '\n'
+const InititalReadmeContents =
+  `# Welcome to GitHub Desktop!${nl}${nl}` +
+  `This is your README. READMEs are where you can communicate ` +
+  `what your project is and how to use it.${nl}${nl}` +
+  `Make any change to this file, save it, and then head ` +
+  `back to GitHub Desktop.${nl}`
+
+/**
+ * A dialog component reponsible for initializing, publishing, and adding
+ * a tutorial repository to the application.
+ */
+export class CreateTutorialRepositoryDialog extends React.Component<
+  ICreateTutorialRepositoryDialogProps,
+  ICreateTutorialRepositoryDialogState
+> {
+  public constructor(props: ICreateTutorialRepositoryDialogProps) {
+    super(props)
+    this.state = { loading: false }
+  }
+
+  private async createAPIRepository(account: Account, name: string) {
+    const api = new API(account.endpoint, account.token)
+
+    try {
+      return await api.createRepository(
+        null,
+        name,
+        'GitHub Desktop tutorial repository',
+        true
+      )
+    } catch (err) {
+      throw new Error(
+        'Could not create the tutorial repository. ' +
+          'The most likely reason is that you already have a repository named ' +
+          `"${name}" on your account at ${friendlyEndpointName(account)}.\n\n` +
+          'Please delete the repository and try again.'
+      )
+    }
+  }
+
+  private async pushRepo(
+    path: string,
+    account: Account,
+    progressCb: (title: string, value: number, description?: string) => void
+  ) {
+    const pushTitle = `Pushing repository to ${friendlyEndpointName(account)}`
+    progressCb(pushTitle, 0)
+
+    const pushOpts = await executionOptionsWithProgress(
+      {
+        env: envForAuthentication(account),
+      },
+      new PushProgressParser(),
+      progress => {
+        if (progress.kind === 'progress') {
+          progressCb(pushTitle, progress.percent, progress.details.text)
+        }
+      }
+    )
+
+    const args = ['push', '-u', 'origin', 'master']
+    await git(args, path, 'tutorial:push', pushOpts)
+  }
+
+  public onSubmit = async () => {
```

Comment:
> perhaps there's some edge cases or complexity we're avoiding with this approach?

No edge cases, just complexity. I don't think we'd be gaining anything by moving this logic to a dispatcher/appstore method since the state contained within this component isn't depended upon or shared with any other part of the application.

Similar patterns can be found in `CreateRepository`, and 'AddExistingRepository' and stands out from for example `PublishRepository` where the progress of the publish operation is shared through app state in order to be communicated to the user through components outside of the publish dialog. The publish operation is also special because it mutates the repository instance itself.

If we were to move this to an App Store method now we'd just be marshalling the progress state through AppState. I also opted for just calling the git methods directly because we don't have a `Repository` instance at this time and changing it such that we did caused more problems than it was worth IMO.